classes/controller/FrontController.php

```diff
@@ -541,31 +541,40 @@ protected function assignGeneralPurposeVariables()
             $cart = new Cart();
         }
 
-        $templateVars = [
-            'cart' => $this->cart_presenter->present($cart),
-            'currency' => $this->getTemplateVarCurrency(),
-            'customer' => $this->getTemplateVarCustomer(),
-            'country' => $this->objectPresenter->present($this->context->country),
-            'language' => $this->objectPresenter->present($this->context->language),
-            'page' => $this->getTemplateVarPage(),
-            'shop' => $this->getTemplateVarShop(),
-            'core_js_public_path' => $this->getCoreJsPublicPath(),
-            'urls' => $this->getTemplateVarUrls(),
-            'configuration' => $this->getTemplateVarConfiguration(),
-            'field_required' => $this->context->customer->validateFieldsRequiredDatabase(),
-            'breadcrumb' => $this->getBreadcrumb(),
-            'link' => $this->context->link,
-            'time' => time(),
-            'static_token' => Tools::getToken(false),
-            'token' => Tools::getToken(),
-            'debug' => _PS_MODE_DEV_,
-        ];
+        $templateVars = [];
+
+        Hook::exec(
+            'actionFrontControllerSetVariablesBefore',
+            [
+                'templateVars' => &$templateVars,
+                'cart' => $cart,
+            ],
+        );
+
+        $templateVars['cart'] = $templateVars['cart'] ?? $this->cart_presenter->present($cart);
+        $templateVars['currency'] = $templateVars['currency'] ?? $this->getTemplateVarCurrency();
+        $templateVars['customer'] = $templateVars['customer'] ?? $this->getTemplateVarCustomer();
```

Comment:
All these assignations are optional: if the variable has already be defined in the hook above, we don't override it.

If the hook was not used, there's no breaking change, we set the same values as before.